src/ucp/rndv/rndv.c

```diff
@@ -1401,34 +1450,33 @@ UCS_PROFILE_FUNC_VOID(ucp_rndv_receive, (worker, rreq, rndv_rts_hdr, rkey_buf),
             }
 
             /* fallback to non get zcopy protocol */
-            ucp_rkey_destroy(rndv_req->send.rndv.rkey);
             is_get_zcopy_failed = 1;
             src_mem_type        = ucp_rkey_packed_mem_type(rkey_buf);
         }
 
-        if (rndv_mode == UCP_RNDV_MODE_AUTO) {
-            /* check if we need pipelined memtype staging */
-            if (UCP_MEM_IS_GPU(rreq->recv.mem_type) &&
-                ucp_rndv_is_recv_pipeline_needed(rndv_req, rndv_rts_hdr,
-                                                 rkey_buf, rreq->recv.mem_type,
-                                                 is_get_zcopy_failed)) {
-                ucp_rndv_recv_data_init(rreq, rndv_rts_hdr->size);
-                if (ucp_rndv_is_put_pipeline_needed(rndv_rts_hdr->address,
-                                                    rndv_rts_hdr->size,
-                                                    get_zcopy->min,
-                                                    get_zcopy->max,
-                                                    is_get_zcopy_failed)) {
-                    /* send FRAG RTR for sender to PUT the fragment. */
-                    ucp_rndv_send_frag_rtr(worker, rndv_req, rreq, rndv_rts_hdr);
-                } else {
-                    /* sender address is present. do GET pipeline */
-                    ucp_rndv_recv_start_get_pipeline(worker, rndv_req, rreq,
-                                                     rndv_rts_hdr->sreq.req_id,
-                                                     rkey_buf,
-                                                     rndv_rts_hdr->address,
-                                                     rndv_rts_hdr->size, 0);
-                }
+        /* Check if pipelined memtype staging is needed */
+        if ((rndv_mode == UCP_RNDV_MODE_AUTO) &&
+            UCP_MEM_IS_GPU(rreq->recv.mem_type) &&
+            ucp_rndv_is_recv_pipeline_needed(rndv_req, rndv_rts_hdr, rkey_buf,
+                                             rreq->recv.mem_type,
+                                             is_get_zcopy_failed)) {
+            put_zcopy = &ep_config->rndv.put_zcopy;
+            ucp_rndv_recv_data_init(rreq, rndv_rts_hdr->size);
+            if (ucp_rndv_is_put_pipeline_needed(rndv_rts_hdr->address,
+                                                rndv_rts_hdr->size, get_zcopy,
+                                                put_zcopy,
+                                                is_get_zcopy_failed)) {
+                /* send FRAG RTR for sender to PUT the fragment. */
+                ucp_rndv_send_frag_rtr(worker, rndv_req, rreq, rndv_rts_hdr);
                 goto out;
+            } else if (is_get_zcopy_supported) {
```

Comment:
isn't presence of remote addr in RTS already saying that get zcopy is supported?

Reply:
it says that sender allows us to use GET Zcopy.
but we need to check local min/max of GET Zcopy to make sure GET Zcopy RNDV is supported by TLs

Reply:
then should we set `is_get_zcopy_failed` to true if we do not get into this `else if`? Because we do not register recv buffer for AM flow

Reply:
@dmitrygx shouldn't above check if (!is_get_zcopy_failed) ?
because is_get_zcopy_supported might be true but is_get_zcopy_failed is false because GDR not available at receiver?

Reply:
> @dmitrygx shouldn't above check if (!is_get_zcopy_failed) ?

I tried it, but it disables pipeline GET Zcopy then.
since we do GET Zcopy first and if it fails (e.g. GDR is not available)m it sets `is_get_zcopy_failed` to `true` and then we should fallback to pipeline GET protocol.

> because is_get_zcopy_supported might be true but is_get_zcopy_failed is false because GDR not available at receiver?

yes, this check aims it. `is_get_zcopy_supported` should be `true` and since we get here it guarantees that `is_get_zcopy_failed` is `true`.